File: 12765 Factorial Products/solution.c
#include <stdio.h>
#include <stdbool.h>

int table[][4] = {
    {}, 
    {}, 
    {1, 0, 0, 0}, 
    {1, 1, 0, 0}, 
    {3, 1, 0, 0}, 
    {3, 1, 1, 0}, 
    {4, 2, 1, 0}, 
    {4, 2, 1, 1}, 
    {7, 2, 1, 1}, 
    {7, 4, 1, 1}
};

bool equal(int *array1, int *array2) {
    for(int i = 0; i < 4; ++i) {
        if(array1[i] != array2[i])
            return false;
    }
    return true;
}

int main() {
    int N, M;
    while(scanf("%d %d", &N, &M) && N != 0 && M != 0) {
        int generator1[4] = {}, generator2[4] = {}, digit;
        while(N--) {
            scanf("%d", &digit);
            for(int i = 0; i < 4; ++i)
                generator1[i] += table[digit][i];
        }
        while(M--) {
            scanf("%d", &digit);
            for(int i = 0; i < 4; ++i)
                generator2[i] += table[digit][i];
        }
        puts(equal(generator1, generator2) ? "YES" : "NO");
    }
    return 0;
}

File: 12765 Factorial Products/solution.java
import java.io.*;
import java.util.*;

public class Main {
    static int[][] table = {
        {0, 0, 0, 0}, 
        {0, 0, 0, 0}, 
        {1, 0, 0, 0}, 
        {1, 1, 0, 0}, 
        {3, 1, 0, 0}, 
        {3, 1, 1, 0}, 
        {4, 2, 1, 0}, 
        {4, 2, 1, 1}, 
        {7, 2, 1, 1}, 
        {7, 4, 1, 1}
    };
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
		StringBuilder output = new StringBuilder();
		while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
		    int N = (int)st.nval;
		    st.nextToken();
		    int M = (int)st.nval;
		    if(N == 0 || M == 0)
		        break;
		    int[] generator1 = new int[4], generator2 = new int[4];
		    while(N-- > 0) {
		        st.nextToken();
                int digit = (int)st.nval;
                for(int i = 0; i < 4; ++i)
                    generator1[i] += table[digit][i];
            }
            while(M-- > 0) {
		        st.nextToken();
                int digit = (int)st.nval;
                for(int i = 0; i < 4; ++i)
                    generator2[i] += table[digit][i];
            }
            output.append(Arrays.equals(generator1, generator2) ? "YES\n" : "NO\n");
		}
		System.out.print(output);
	}
}
